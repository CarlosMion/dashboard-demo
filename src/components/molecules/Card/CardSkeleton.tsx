import React from "react";
import Skeleton from "@mui/material/Skeleton";
import { styled } from "@mui/material/styles";
import Grid from "@mui/material/Grid2";
import useMediaQuery from "@mui/material/useMediaQuery";

const CardSkeletonContainer = styled("div")(({ theme }) => ({
  padding: theme.spacing(2),
  borderRadius: theme.shape.borderRadius,
  boxShadow: theme.shadows[1],
  backgroundColor: theme.palette.background.paper,
  width: useMediaQuery(theme.breakpoints.down("md")) ? "265px" : "350px",
  margin: "auto",
}));

const CardSkeleton = () => {
  return (
    <CardSkeletonContainer>
      <Grid container spacing={2}>
        <Grid size={12}>
          <Skeleton variant="text" width="60%" height={30} />
        </Grid>
        <Grid size={6}>
          <Skeleton variant="text" width="80%" height={20} />
        </Grid>
        <Grid size={6}>
          <Skeleton variant="text" width="80%" height={20} />
        </Grid>
        <Grid size={12}>
          <Skeleton variant="text" width="100%" height={30} />
        </Grid>
        <Grid size={12}>
          <Skeleton variant="rectangular" width="100%" height={40} />
        </Grid>
      </Grid>
    </CardSkeletonContainer>
  );
};

export default CardSkeleton;
