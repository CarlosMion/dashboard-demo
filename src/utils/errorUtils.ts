export type ServerError = {
  error: string;
  statusCode?: number;
};

export const isServerErrorWithStatusCode = (
  result: object
): result is ServerError => {
  return "error" in result && "statusCode" in result;
};

export const isServerError = (result: unknown): result is ServerError => {
  return typeof result === "object" && result !== null && "error" in result;
};

type ErrorWithMessage = {
  message: string;
};

export const isErrorWithMessage = (
  error: unknown
): error is ErrorWithMessage => {
  return (
    typeof error === "object" &&
    error !== null &&
    "message" in error &&
    typeof (error as Record<string, unknown>).message === "string"
  );
};

export const toErrorWithMessage = (maybeError: unknown): ErrorWithMessage => {
  if (isErrorWithMessage(maybeError)) return maybeError;

  try {
    return new Error(JSON.stringify(maybeError));
  } catch {
    // fallback in case there's an error stringifying the maybeError
    // like with circular references for example.
    return new Error(String(maybeError));
  }
};

export const getErrorMessage = (error: unknown): string => {
  return toErrorWithMessage(error).message;
};

export class ApiError extends Error {
  statusCode: number;

  constructor(message: string, statusCode: number) {
    super(message);
    this.name = "ApiError";
    this.statusCode = statusCode || 500;
  }
}
