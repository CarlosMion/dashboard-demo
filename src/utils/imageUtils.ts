export const dataURItoBlob = (dataURI: string, fileType: string): Blob => {
  const byteString = atob(dataURI.split(',')[1]);
  const arrayBuffer = new ArrayBuffer(byteString.length);
  const intArray = new Uint8Array(arrayBuffer);

  for (let i = 0; i < byteString.length; i++) {
    intArray[i] = byteString.charCodeAt(i);
  }

  return new Blob([arrayBuffer], { type: fileType });
};

export const dataURItoFile = (
  dataURI: string,
  fileName: string,
  fileType: string
): File => {
  const blob = dataURItoBlob(dataURI, fileType);
  return new File([blob], fileName, { type: blob.type });
};

export const blobToFile = (blob: Blob, fileName: string): File => {
  return new File([blob], fileName, { type: blob.type });
};

export const getImageDimensions = (
  src: string
): Promise<{ width: number; height: number } | null> => {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.src = src;

    img.onload = () =>
      resolve(
        img.width && img.height
          ? { width: img.width, height: img.height }
          : null
      );
    img.onerror = (error) => reject(error);
  });
};

export const isImageDimensionWithinThreshold = (
  actual: number,
  ideal: number,
  threshold: number
): boolean => {
  const isWithinThreshold = actual >= ideal * (1 - threshold);
  return isWithinThreshold;
};
