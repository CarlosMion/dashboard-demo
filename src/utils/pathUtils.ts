export function toCamelCase(str?: string): string | undefined {
  return str?.replace(/-([a-z])/g, (match, letter) => letter.toUpperCase());
}
