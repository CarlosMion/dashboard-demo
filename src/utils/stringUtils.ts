export function maskCardNumber(cardNumber: number): string {
  const cardNumberStr = cardNumber.toString();
  return cardNumberStr.replace(/(\d{4})\d{8}(\d{4})/, "$1 **** **** $2");
}

export function formatMoney(value: number): string {
  return value.toLocaleString("en-US", { style: "currency", currency: "USD" });
}

export function formatValidThru(value: string): string {
  return new Date(value).toLocaleDateString("en-US", {
    month: "2-digit",
    year: "2-digit",
    timeZone: "UTC",
  });
}

export function formatTransactionDate(value: string): string {
  return new Date(value).toLocaleDateString("en-US", {
    day: "2-digit",
    month: "long",
    year: "numeric",
    timeZone: "UTC",
  });
}
