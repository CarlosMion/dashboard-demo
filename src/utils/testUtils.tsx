import { render } from "@testing-library/react";
import { ReactNode } from "react";
import AppProviders from "@/providers/AppProviders/AppProviders";

import messages from "@/i18n/messages/en.json";

const customRender = (children: ReactNode) =>
  render(
    <AppProviders
      locale={"en"}
      messages={messages}
      userAgent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
    >
      {children}
    </AppProviders>
  );

export * from "@testing-library/react";

export { customRender as render };
