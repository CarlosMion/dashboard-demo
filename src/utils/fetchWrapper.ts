import { isServerError } from "./errorUtils";

const fetcher = async <T>(
  info: RequestInfo | URL,
  init?: RequestInit
): Promise<T> => {
  let result;
  try {
    const injectedHeaders: Record<string, string> = {};
    if (typeof window === "undefined") {
      const headers = await import("next/headers");
      (await headers.headers()).forEach((value, key) => {
        injectedHeaders[key] = value;
      });
    }

    const buildHeaders: Record<string, string> = {};
    let url = info;
    if (typeof info === "string") {
      url = `${process.env.NEXT_PUBLIC_WEBAPP_API_URL}/${info}`;
    }
    const reqHeaders = {
      "Content-Type": "application/json",
      ...init?.headers,
      ...injectedHeaders,
      ...buildHeaders,
    };
    result = await fetch(url, {
      ...init,
      headers: reqHeaders,
    });
    if (!result.ok) {
      throw new Error("Failed to fetch");
    }
    let parsedResult;
    try {
      parsedResult = await result.json();
    } catch (error) {
      console.log("Failed to parse response, probably not JSON", result, error);
      return result as T;
    }
    if (isServerError(parsedResult)) {
      throw new Error(parsedResult.error);
    }
    return parsedResult as T;
  } catch (error) {
    throw error;
  }
};

export default fetcher;
