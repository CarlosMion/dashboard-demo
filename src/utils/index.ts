import { isEdge } from "./userAgentUtils";
import { maskCardNumber } from "./stringUtils";
import { toCamelCase } from "./pathUtils";
import fetcher from "./fetchWrapper";
import {
  isServerErrorWithStatusCode,
  ApiError,
  getErrorMessage,
  isErrorWithMessage,
  isServerError,
  toErrorWithMessage,
} from "./errorUtils";
import {
  blobToFile,
  dataURItoBlob,
  dataURItoFile,
  getImageDimensions,
  isImageDimensionWithinThreshold,
} from "./imageUtils";
import {
  createImage,
  getRadianAngle,
  rotateSize,
  getCroppedImg,
} from "./cropImageUtils";
export {
  fetcher,
  isEdge,
  toCamelCase,
  isServerErrorWithStatusCode,
  ApiError,
  getErrorMessage,
  isErrorWithMessage,
  isServerError,
  toErrorWithMessage,
  blobToFile,
  dataURItoBlob,
  dataURItoFile,
  getImageDimensions,
  isImageDimensionWithinThreshold,
  getCroppedImg,
  rotateSize,
  getRadianAngle,
  createImage,
  maskCardNumber,
};
