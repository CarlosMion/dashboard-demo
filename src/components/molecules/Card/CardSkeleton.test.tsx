import { render } from "@/utils/testUtils";
import CardSkeleton from "./CardSkeleton";

describe("CardSkeleton", () => {
  it("renders correctly and matches snapshot", () => {
    const { container } = render(<CardSkeleton />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders all skeleton elements", () => {
    const { container } = render(<CardSkeleton />);
    const skeletons = container.querySelectorAll(".MuiSkeleton-root");
    expect(skeletons).toHaveLength(8);
  });

  it("renders with correct variant", () => {
    const { container } = render(<CardSkeleton />);
    expect(container.firstChild).toHaveAttribute("variant", "dark");
  });

  it("applies correct styles to container", () => {
    const { container } = render(<CardSkeleton />);
    const gridContainer = container.firstChild;
    expect(gridContainer).toHaveClass("MuiGrid2-root");
    expect(gridContainer).toHaveClass("MuiGrid2-container");
  });
});
