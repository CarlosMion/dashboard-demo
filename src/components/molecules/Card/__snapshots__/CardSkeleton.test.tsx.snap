// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardSkeleton renders correctly and matches snapshot 1`] = `
<div
  class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row css-i1p3el-MuiGrid2-root"
  variant="dark"
>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-9.5 css-bcxyza-MuiGrid2-root"
  >
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1mvk69b-MuiSkeleton-root"
      style="width: 100px; height: 20px;"
    />
    <span
      class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-w8k4ky-MuiSkeleton-root"
      style="width: 100%; height: 15px;"
    />
  </div>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-2.5 css-1ykqhke-MuiGrid2-root"
  >
    <span
      class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-pulse css-26dym8-MuiSkeleton-root"
      style="width: 40px; height: 30px;"
    />
  </div>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-5 css-ufm42q-MuiGrid2-root"
  >
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1mvk69b-MuiSkeleton-root"
      style="width: 100px; height: 20px;"
    />
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1mvk69b-MuiSkeleton-root"
      style="width: 80%; height: 20px;"
    />
  </div>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-5 css-uytjhf-MuiGrid2-root"
  >
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1mvk69b-MuiSkeleton-root"
      style="width: 100px; height: 20px;"
    />
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1mvk69b-MuiSkeleton-root"
      style="width: 80%; height: 20px;"
    />
  </div>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 css-1j0u08r-MuiGrid2-root"
  >
    <span
      class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-pr39we-MuiSkeleton-root"
      style="width: 100%; height: 30px;"
    />
  </div>
</div>
`;
